File: py/lc/767_reorganize_string.py
"""
https://leetcode.com/problems/reorganize-string/?envType=company&envId=roblox&favoriteSlug=roblox-all

Given a string s, rearrange the characters of s so that any two adjacent characters are not the same.

Return any possible rearrangement of s or return "" if not possible.

Example 1:
Input: s = "aab"
Output: "aba"

Example 2:
Input: s = "aaab"
Output: ""

Constraints:
1 <= s.length <= 500
s consists of lowercase English letters.
"""

from collections import Counter


class Solution:
    def reorganize_string(self, s: str) -> str:
        char_count = Counter(s)
        max_count, letter = 0, ""

        for char, count in char_count.items():
            if count >= max_count:
                max_count = count
                letter = char

        if max_count > (len(s) + 1) // 2:
            return ""

        result = [" "] * len(s)
        idx = 0

        while char_count[letter] != 0:
            result[idx] = letter
            idx += 2
            char_count[letter] -= 1

        for char, count in char_count.items():
            while count > 0:
                if idx >= len(s):
                    idx = 1

                result[idx] = char
                idx += 2
                count -= 1

        return "".join(result)


def test_one():
    solution = Solution()
    assert solution.reorganize_string("aab") == "aba"


def test_two():
    solution = Solution()
    assert solution.reorganize_string("aaab") == ""

File: rs/src/lc/767_reorganize_string.rs
use std::collections::HashMap;

fn reorganize_string(s: String) -> String {
    let n = s.chars().count();
    if n <= 1 {
        return s.to_string();
    }

    let mut counts = s.chars().fold(HashMap::new(), |mut m, c| {
        *m.entry(c).or_insert(0) += 1;
        m
    });

    let (max_count, letter) = counts
        .iter()
        .fold((0usize, '\0'), |(max_c, max_ch), (&ch, &cnt)| {
            if cnt >= max_c {
                (cnt, ch)
            } else {
                (max_c, max_ch)
            }
        });

    if max_count > (n + 1) / 2 {
        return String::new();
    }

    let mut res = std::iter::repeat(' ').take(n).collect::<Vec<_>>();
    let mut idx = 0usize;

    let cnt = counts.get_mut(&letter).unwrap();
    while *cnt > 0 {
        res[idx] = letter;
        idx += 2;
        *cnt -= 1;
    }

    for (ch, mut cnt) in counts.into_iter() {
        while cnt > 0 {
            if idx >= n {
                idx = 1;
            }
            res[idx] = ch;
            idx += 2;
            cnt -= 1;
        }
    }

    res.into_iter().collect()
}

#[test]
fn test_one() {
    let result = reorganize_string(String::from("aab"));
    assert_eq!(result, "aba")
}

#[test]
fn test_two() {
    let result = reorganize_string(String::from("aaab"));
    assert_eq!(result, "");
}
